mitodl/micromasters
# Refactor user edx data fetching

#### What are the relevant tickets?

No issue - done as a precursor to #933 
**Rebased on #1890 (to help with testing)**

#### What's this PR do?

Streamlines user edX data fetching and avoids potential bugs 

#### Where should the reviewer start?

dashboard/api_edx_cache.py and dashboard/models.py

Most of these changes are naming changes and repetitive changes to parameters, etc. The most significant functional change is the addition of [CachedEdxUserData](https://github.com/mitodl/micromasters/compare/master...refactor_user_edx_data_fetching#diff-54f823943ee528710d6f9d178443a994L18) and 

#### How should this be manually tested?

To test that the dashboard API still works correctly:

1. Unseed and seed the database, specifying `--staff-user=YOUR_USERNAME` when you run `seed_db`
1. Check the /dashboard page and make sure 'Digital Learning' and 'Analog Learning' look right
1. Enroll in a non-FA course using the `alter_data` commands. Example: `manage.py alter_data --action=set_course_to_enrolled --username=YOUR_USERNAME --program-title='Analog' --course-title='100'`
1. Reload the /dashboard page and check that the enrolled course shows up correctly

To test that serialization for search results still works:

1. Look at the /learners page and pick one of the fake 'Analog Learning' program users
1. Set one or more courses to passed for that user. Example: `alter_data --action=set_course_to_passed --email=roberto.cortes --program-title='Analog' --course-title='100' --grade=100`
1. Run recreate_index: `manage.py recreate_index`
1. Reload the page and confirm that the user's average grade has been changed
